src/net.cpp

```diff
@@ -449,102 +460,108 @@ CNode* CConnman::ConnectNode(CAddress addrConnect, const char *pszDest, bool fCo
+                            m_unused_i2p_sessions.pop();
+                        }
                     }
-                }
-                connected = i2p_transient_session->Connect(addrConnect, conn, proxyConnectionFailed);
-                if (!connected) {
-                    LOCK(m_unused_i2p_sessions_mutex);
-                    if (m_unused_i2p_sessions.size() < MAX_UNUSED_I2P_SESSIONS_SIZE) {
-                        m_unused_i2p_sessions.emplace(i2p_transient_session.release());
+                    connected = i2p_transient_session->Connect(target_addr, conn, proxyConnectionFailed);
+                    if (!connected) {
+                        LOCK(m_unused_i2p_sessions_mutex);
+                        if (m_unused_i2p_sessions.size() < MAX_UNUSED_I2P_SESSIONS_SIZE) {
+                            m_unused_i2p_sessions.emplace(i2p_transient_session.release());
+                        }
                     }
                 }
-            }
 
-            if (connected) {
-                sock = std::move(conn.sock);
-                addr_bind = CAddress{conn.me, NODE_NONE};
+                if (connected) {
+                    sock = std::move(conn.sock);
+                    addr_bind = CAddress{conn.me, NODE_NONE};
+                }
+            } else if (use_proxy) {
+                sock = CreateSock(proxy.proxy);
+                if (!sock) {
+                    return nullptr;
+                }
+                connected = ConnectThroughProxy(proxy, target_addr.ToStringAddr(), target_addr.GetPort(),
+                                                *sock, nConnectTimeout, proxyConnectionFailed);
+            } else {
+                // no proxy needed (none set for target network)
+                sock = CreateSock(target_addr);
+                if (!sock) {
+                    continue;
+                }
+                connected = ConnectSocketDirectly(target_addr, *sock, nConnectTimeout,
+                                                conn_type == ConnectionType::MANUAL);
             }
-        } else if (use_proxy) {
-            sock = CreateSock(proxy.proxy);
-            if (!sock) {
-                return nullptr;
+            if (!proxyConnectionFailed) {
+                // If a connection to the node was attempted, and failure (if any) is not caused by a problem connecting to
+                // the proxy, mark this as an attempt.
+                addrman.Attempt(target_addr, fCountFailure);
             }
-            connected = ConnectThroughProxy(proxy, addrConnect.ToStringAddr(), addrConnect.GetPort(),
-                                            *sock, nConnectTimeout, proxyConnectionFailed);
-        } else {
-            // no proxy needed (none set for target network)
-            sock = CreateSock(addrConnect);
+        } else if (pszDest && GetNameProxy(proxy)) {
+            sock = CreateSock(proxy.proxy);
             if (!sock) {
                 return nullptr;
```

Comment:
nit: can you add a comment to justify early termination here (instead of trying going through other addrs)